### ISSUE TITLE
Add support for Date and Datetime in scalar indexes via UNIX integer representation of time?

### ISSUE THREAD

**Issue by mkleinbort-ic** [lancedb/lance#1945]
Very happy with lance! It's a great data format.

However, it's cumbersome for me to have to save columns `eventDate` and `eventDateInt` (an integer version of the same) just to get indexing on the event date.

I think Lance could easily offer a near-invisible conversion of Data/Datetime to UNIX integers.

**Comment by changhiskhan** [comment]
thanks!

What's the data type for the `eventDate` column? @westonpace already added support for scalar indices on datetime columns. Does it not work for you?

**Comment by westonpace** [comment]
Yeah, I was thinking this would work but it does look like we restricted it:

```
TypeError: Scalar index column instants must be int, float, bool, or str
```

We store all temporal types as integers anyways and datafusion has support for comparing them so this should be straightforward.  I tried disabling the check and it looks like there are a few TODOs remaining.  The first one I ran into is that we need to add temporal types to `lance_datafusion::expr::safe_coerce_scalar` (it looks like all timestamp literals are converted to microseconds but we should be able to safely cast to any other resolution).
